File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;
import java.util.stream.*;


class Solution {
    public int solution(int[][] js) {
        Arrays.sort(js, Comparator.comparing((int[] job) -> job[0]).thenComparing(job -> job[1]));
        
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparing((int[] job) -> job[1]).thenComparing(job -> job[0]));
        pq.add(js[0]);
        int index = 1;
        int currentTime = js[0][0];
        int answer = 0;
        
        while (!pq.isEmpty()) {
            int[] job = pq.poll();
           
            
            currentTime += job[1];
            answer += currentTime - job[0];
 
            while (index < js.length && js[index][0] <= currentTime) {
                pq.add(js[index]);
                index++;
            }
            
            if (pq.isEmpty() && index < js.length) {
                pq.add(js[index]);
                currentTime = js[index][0];
                index++;
            }
        }
        
        return answer / js.length;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
from collections import deque
from heapq import heappop, heappush

def solution(jobs):
    jobs.sort(key = lambda x : (x[0], x[1]))
    q = [ [jobs[0][1], jobs[0][0]] ] 
    index = 1
    current_time = jobs[0][0]
    answer = 0
    
    while q:
        duration, req_time = heappop(q)
        current_time += duration
        answer += current_time - req_time
        
        while index < len(jobs) and jobs[index][0] <= current_time:
            heappush(q, [jobs[index][1], jobs[index][0]])
            index += 1
        
        if not q and index < len(jobs):
            heappush(q, [jobs[index][1], jobs[index][0]])
            current_time = jobs[index][0]
            index += 1
        
    return answer // len(jobs)
            
            
    
        
        

        